File: 백준/Gold/1916. 최소비용 구하기/최소비용 구하기.py
import sys
import heapq

input = sys.stdin.readline
hpush = heapq.heappush
hpop = heapq.heappop

n = int(input())
m = int(input())

q = []
bus = [[] for _ in range(n+1)]
dist = [2147483647] * (n+1)

for _ in range(m):
    s, t, d = map(int, input().split(' '))
    bus[s].append((d, t))

s, t = map(int, input().split(' '))
dist[s] = 0
hpush(q, (0, s))

while len(q) != 0:
    cur_dist, cur = hpop(q)

    if cur_dist > dist[cur]:
        continue

    for nxt_dist, nxt in bus[cur]:
        nxt_dist = nxt_dist + cur_dist
        if nxt_dist < dist[nxt]:
            hpush(q, (nxt_dist, nxt))
            dist[nxt] = nxt_dist

print(dist[t])

File: 백준/Gold/1916. 최소비용 구하기/최소비용 구하기.rs
use std::io::{stdin, stdout, Read, Write, BufWriter};
use std::collections::{binary_heap, BinaryHeap};
use std::cmp::Ordering;

#[derive(Clone, Copy, PartialEq, Eq)]
struct Edge {
    cost: usize,
    node: usize,
}

impl Ord for Edge {
    fn cmp(&self, other: &Self) -> Ordering {
        other.cost.cmp(&self.cost)
            .then_with(|| self.node.cmp(&other.node))
    }
}

impl PartialOrd for Edge {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn main(){
    let mut input = String::new();
    stdin().read_to_string(&mut input).unwrap();
    let mut input = input.split_ascii_whitespace().flat_map(str::parse::<usize>);

    let (n, m) = (input.next().unwrap(), input.next().unwrap());
    let mut dist = vec![2147483647; n+1];
    let mut bus: Vec<Vec<Edge>> = vec![vec![]; n+1];

    for _ in 0..m {
        let (s, t, d) = (input.next().unwrap(), input.next().unwrap(), input.next().unwrap());
        bus[s].push(Edge { cost: d, node: t })
    }

    let (s, t) = (input.next().unwrap(), input.next().unwrap());
    let mut pq = BinaryHeap::new();
    pq.push(Edge { cost: 0, node: s});
    dist[s] = 0;

    while !pq.is_empty() {
        let cur = pq.pop().unwrap();
        if cur.cost > dist[cur.node] {
            continue;
        }

        for nxt in bus[cur.node].iter() {
            if dist[nxt.node] > nxt.cost+cur.cost {
                pq.push(Edge { cost: nxt.cost+cur.cost, node: nxt.node });
                dist[nxt.node] = nxt.cost+cur.cost;
            }
        }
    }

    let stdout = stdout();
    let mut writer = BufWriter::new(stdout);

    writeln!(writer, "{}", dist[t]).unwrap();

} 